File: notifications/src/main/java/io/castled/notifications/trigger/NumberConditionEvaluator.java
package io.castled.notifications.trigger;

import java.text.NumberFormat;
import java.text.ParseException;

import io.castled.notifications.exceptions.CastledRuntimeException;
import io.castled.notifications.trigger.models.SingleValueOperation;
import io.castled.notifications.trigger.models.BetweenOperation;
import io.castled.notifications.trigger.models.PropertyOperation;

public class NumberConditionEvaluator implements ParamsConditionEvaluator {
    @Override
    public boolean evaluateCondition(Object value, PropertyOperation propertyOperation) {
        Number numberValue = (Number) value;
        switch (propertyOperation.getOperationType()) {
            case EQ:
                Number conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return numberValue != null && conditionValue.longValue() == numberValue.longValue();
            case NEQ:
                if (numberValue == null) {
                    return true;
                }
                conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return conditionValue.longValue() != numberValue.longValue();
            case GT:
                conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return numberValue != null && conditionValue.doubleValue() < numberValue.doubleValue();
            case LT:
                conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return numberValue != null && conditionValue.doubleValue() > numberValue.doubleValue();
            case GTE:
                conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return numberValue != null && conditionValue.longValue() <= numberValue.longValue();
            case LTE:
                conditionValue = getConditionValue(((SingleValueOperation) propertyOperation).getValue());
                return numberValue != null && conditionValue.longValue() >= numberValue.longValue();
            case BETWEEN:
                BetweenOperation doubleValueOperation = ((BetweenOperation) propertyOperation);
                Number fromValue = getConditionValue(doubleValueOperation.getFrom());
                Number toValue = getConditionValue(doubleValueOperation.getTo());
                return numberValue != null && fromValue.doubleValue() < numberValue.doubleValue() && toValue.doubleValue() > numberValue.doubleValue();
            default:
                throw new CastledRuntimeException(String.format("Operations type %s not supported for numeric operand", propertyOperation.getPropertyType()));
        }
    }

    private Number getConditionValue(String value) {
        try {
            return NumberFormat.getInstance().parse(value);
        } catch (ParseException e) {
            throw new CastledRuntimeException(e);
        }
    }
}

File: notifications/src/main/java/io/castled/notifications/trigger/NumberConditionEvaluator.kt
package io.castled.notifications.trigger

import io.castled.notifications.trigger.enums.OperationType
import io.castled.notifications.trigger.models.BetweenOperation
import io.castled.notifications.trigger.models.PropertyOperation
import io.castled.notifications.trigger.models.SingleValueOperation
import java.text.NumberFormat
import java.text.ParseException

internal class NumberConditionEvaluator : ParamsConditionEvaluator() {
    override fun evaluateCondition(value: Any?, propertyOperation: PropertyOperation): Boolean {
        val paramValue = value as Number? ?: return false
        when (propertyOperation.operationType) {
            OperationType.EQ -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { it.toLong() == paramValue.toLong() } ?: false
            }
            OperationType.NEQ -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { it.toLong() != paramValue.toLong() } ?: false
            }
            OperationType.GT -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { paramValue.toLong() > it.toLong() } ?: false
            }
            OperationType.LT -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { paramValue.toLong() < it.toLong() } ?: false
            }
            OperationType.GTE -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { paramValue.toLong() >= it.toLong() } ?: false
            }
            OperationType.LTE -> {
                val conditionValue =
                    getConditionValue((propertyOperation as SingleValueOperation).value)
                return conditionValue?.let { paramValue.toLong() <= it.toLong() } ?: false
            }
            OperationType.BETWEEN -> {
                val doubleValueOperation = propertyOperation as BetweenOperation
                val fromValue = getConditionValue(doubleValueOperation.from)
                val toValue = getConditionValue(doubleValueOperation.to)
                return if (fromValue != null && toValue != null)
                    paramValue.toLong() >= fromValue.toLong() && paramValue.toLong() <= toValue.toLong()
                else false
            }
            else -> {
                logger.error("Operations type ${propertyOperation.propertyType} not supported for numeric operand")
                return false
            }
        }
    }

    private fun getConditionValue(value: String): Number? {
        return try {
            NumberFormat.getInstance().parse(value)
        } catch (e: ParseException) {
            return null
        }
    }
}